src/script/sign.h

```diff
@@ -68,6 +71,432 @@ struct SignatureData {
+                        redeemscript_bytes.resize(value_len);
+                        Span<unsigned char> redeemscript_data = MakeSpan(redeemscript_bytes);
+                        s >> redeemscript_data;
+                        CScript redeemscript(redeemscript_bytes.begin(), redeemscript_bytes.end());
+
+                        // Check that the redeemscript hash160 matches the one provided
+                        hash160_hasher.Write(&redeemscript_bytes[0], redeemscript_bytes.size());
+                        unsigned char rs_hash160_data[hash160_hasher.OUTPUT_SIZE];
+                        hash160_hasher.Finalize(rs_hash160_data);
+                        uint160 rs_hash160(std::vector<unsigned char>(rs_hash160_data, rs_hash160_data + hash160_hasher.OUTPUT_SIZE));
+                        if (hash160.Compare(rs_hash160) != 0) {
+                            throw std::ios_base::failure("Provided hash160 does not match the redeemscript's hash160");
+                        }
+
+                        // Add to map
+                        redeem_scripts.emplace(hash160, redeemscript);
+                    } else {
+                        // Read in the utxo
+                        CTxOut vout;
+                        s >> vout;
+
+                        // Add to map
+                        input.witness_utxo = vout;
+                    }
+                    break;
+                // witnessscript or a partial signature
+                case PSBT_WITNESSSCRIPT_PARTIAL_SIG:
+                    if (in_globals) {
+                        // Make sure that the key is the size of uint256 + 1
+                        if (key.size() != 33) {
+                            throw std::ios_base::failure("Size of key was not the expected size for the type witness script");
+                        }
+                        // Retrieve sha256 from key
+                        uint256 hash(std::vector<unsigned char>(key.begin() + 1, key.end()));
+
+                        // Read in the redeemscript
+                        std::vector<unsigned char> witnessscript_bytes;
+                        witnessscript_bytes.resize(value_len);
+                        Span<unsigned char> witnessscript_data = MakeSpan(witnessscript_bytes);
+                        s >> witnessscript_data;
+                        CScript witnessscript(witnessscript_bytes.begin(), witnessscript_bytes.end());
+
+                        // Check that the witnessscript sha256 matches the one provided
+                        sha256_hasher.Write(&witnessscript_bytes[0], witnessscript_bytes.size());
+                        unsigned char ws_sha256_data[sha256_hasher.OUTPUT_SIZE];
+                        sha256_hasher.Finalize(ws_sha256_data);
+                        uint256 ws_sha256(std::vector<unsigned char>(ws_sha256_data, ws_sha256_data + sha256_hasher.OUTPUT_SIZE));
+                        if (hash.Compare(ws_sha256) != 0) {
```

Comment:
In commit "Implement PSBT structures and un/serliazation methods per BIP 174":

Same here about `(hash != ws_sha256)`.

Reply:
Done